samples/drivers/flash_shell/src/main.c

```diff
@@ -160,30 +161,35 @@ static int parse_u8(const char *str, uint8_t *result)
 }
 
 /* Read bytes, dumping contents to console and printing on error. */
-static int do_read(const struct shell *sh, off_t offset, size_t len)
+static int do_read(const struct shell *sh, off_t offset, size_t len,
+		   uint8_t *cmp_buf)
 {
 	uint8_t buf[64];
 	int ret;
-
-	while (len > sizeof(buf)) {
-		ret = flash_read(flash_device, offset, buf, sizeof(buf));
-		if (ret) {
-			goto err_read;
+	size_t read_len;
+	bool cmp_ok = true;
+
+	do {
+		read_len = len > sizeof(buf) ? sizeof(buf) : len;
+		ret = flash_read(flash_device, offset, buf, read_len);
+		if (ret != 0) {
+			PR_ERROR(sh, "flash_read error: %d\n", ret);
+			return ret;
 		}
-		dump_buffer(sh, buf, sizeof(buf));
-		len -= sizeof(buf);
-		offset += sizeof(buf);
-	}
-	ret = flash_read(flash_device, offset, buf, len);
-	if (ret) {
-		goto err_read;
+		dump_buffer(sh, buf, read_len);
+		if (cmp_buf != NULL) {
+			cmp_ok &= memcmp(cmp_buf, buf, read_len) == 0;
+			cmp_buf += read_len;
+		}
```

Comment:
```
if (cmp_ok && cmp_buf !=NULL) {
  cmp_ok = ...
}
```
if `cmp_ok` gets false, we are no longer interested in comparing anything, we already know buffer is bad.
Comparing every buffer only makes sense if you are going to notify failure separately for each buffer.

Reply:
That makes sense. I have tried to add you suggestion in a [separate commit](https://github.com/zephyrproject-rtos/zephyr/pull/53857/commits/f903901e1eaa90f570d4db6fd14ccb812798d61a). Here is a sample output:
![image](https://user-images.githubusercontent.com/4233729/233992232-bd2bb312-dfee-485d-bef6-83e94bc73c52.png)

Reply:
Note that I modified the write_pattern locally to make the read back fail.

Reply:
> Note that I modified the write_pattern locally to make the read back fail.

Looks nice. I would also print the "Write verification error, unexpected values marked red" or something like that, because it is not clear why there are red

Reply:
That's a good idea. I've changed the commit a bit.